app/scripts/controllers/network/network-controller.test.js

```diff
@@ -50,224 +69,4106 @@ const PRE_1559_BLOCK = {
+            state: {
+              provider: {
+                type: 'rpc',
+                rpcUrl: 'https://mock-rpc-url',
+                chainId: '0x1337',
+              },
+            },
+          },
+          async ({ controller, network }) => {
+            network.mockEssentialRpcCalls();
+
+            const infuraIsUnblocked = await waitForEvent({
+              controller,
+              eventName: 'infuraIsUnblocked',
+              operation: async () => {
+                await controller.initializeProvider();
+              },
+            });
+
+            expect(infuraIsUnblocked).toBe(true);
+          },
+        );
+      });
+
+      it('persists the network version to state (assuming that the request for net_version responds successfully)', async () => {
+        await withController(
+          {
+            state: {
+              provider: {
+                type: 'rpc',
+                rpcUrl: 'https://mock-rpc-url',
+                chainId: '0x1337',
+              },
+            },
+          },
+          async ({ controller, network }) => {
+            network.mockEssentialRpcCalls({
+              net_version: {
+                response: {
+                  result: '42',
+                },
+              },
+            });
+
+            await controller.initializeProvider();
+
+            expect(controller.store.getState().network).toStrictEqual('42');
+          },
+        );
+      });
+
+      it('persists to state whether the network supports EIP-1559 (assuming that the request for eth_getBlockByNumber responds successfully)', async () => {
+        await withController(
+          {
+            state: {
+              provider: {
+                type: 'rpc',
+                rpcUrl: 'https://mock-rpc-url',
+                chainId: '0x1337',
+              },
+              networkDetails: {
+                EIPS: {},
+              },
+            },
+          },
+          async ({ controller, network }) => {
+            network.mockEssentialRpcCalls({
+              latestBlock: POST_1559_BLOCK,
+            });
+
+            await controller.initializeProvider();
+
+            expect(
+              controller.store.getState().networkDetails.EIPS['1559'],
+            ).toBe(true);
+          },
+        );
+      });
     });
+  });
 
-    describe('#getEIP1559Compatibility', () => {
-      it('should return false when baseFeePerGas is not in the block header', async () => {
-        await withController(async ({ controller }) => {
-          setupMockRpcBlockResponses({ block: PRE_1559_BLOCK });
-          await controller.initializeProvider();
-          const supportsEIP1559 = await controller.getEIP1559Compatibility();
-          expect(supportsEIP1559).toStrictEqual(false);
-        });
+  describe('getProviderAndBlockTracker', () => {
+    it('returns objects that proxy to the provider and block tracker as long as the provider has been initialized', async () => {
```

Comment:
Nit: It would be great to have a test that shows that these always point at the current network, even after the network switches.

Reply:
Ah okay. That must have been what you meant in your TODOs here. So just to clarify, you want a test that initializes the provider, grabs the chain ID of the current network using the provider proxy, switches the network, grabs the chain ID of the new network using the provider proxy, and asserts that they are different?

Reply:
Yeah exactly!

Reply:
Okay, tests added here: [9f4e7d0](https://github.com/MetaMask/metamask-extension/pull/17498/commits/9f4e7d0633ba44241c4a3c88ee7403001e302dde)

Reply:
Hmm, it looks like these tests aren't covering the proxy behavior at the moment. I think you'd want to check the chain ID before and after switching chains, of the same provider instance. That would confirm that that single instance is being updated via proxy to the new chain.

Reply:
Ah — you're right, thanks for pointing that out. I've been staring at this too long, I guess. Fixed here: [8b05ade](https://github.com/MetaMask/metamask-extension/pull/17498/commits/8b05adef026ffe0bead287c81f0c06ac1717089f)